File: api/core/hunt/ingestion/tools/crtsh.py
"""crt.sh — certificate transparency log search.

No API key. Returns every certificate ever issued for a domain,
which is gold for subdomain enumeration and pivot-by-cert-fingerprint.
"""
from __future__ import annotations

import json
from typing import Any, List, Mapping

import httpx

from hunt.ingestion.base import Finding, ToolFunction
from hunt.models import EntityKind, SourceTool


class CrtshTool(ToolFunction):
    name = SourceTool.CRT_SH
    accepts = {"domain"}
    emits = {EntityKind.SUBDOMAIN, EntityKind.CERT, EntityKind.ORG}
    description = "Certificate Transparency log search via crt.sh."
    docs_url = "https://crt.sh/"
    per_request_timeout = 20.0

    URL = "https://crt.sh/"

    def _fetch(self, query: Mapping[str, Any]) -> httpx.Response:
        target = query["target"].strip().lower()
        with httpx.Client(timeout=self.per_request_timeout) as client:
            return client.get(
                self.URL,
                params={"q": f"%.{target}", "output": "json"},
            )

    def _parse(self, raw: bytes, query: Mapping[str, Any]) -> List[Finding]:
        try:
            rows = json.loads(raw.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            return []
        if not isinstance(rows, list):
            return []
        findings: List[Finding] = []
        target = query["target"].strip().lower()
        seen_sub: set[str] = set()
        for row in rows[:200]:
            name = (row.get("name_value") or "").strip().lower()
            if not name:
                continue
            for sub in name.split("\n"):
                sub = sub.strip()
                if not sub.endswith(target) and sub != target:
                    continue
                if sub not in seen_sub:
                    seen_sub.add(sub)
                    findings.append(
                        Finding(
                            source_tool=self.name,
                            entity_kind=EntityKind.SUBDOMAIN,
                            entity_value=sub,
                            attributes={"parent_domain": target},
                        )
                    )
            cert_id = row.get("id")
            issuer = (row.get("issuer_name") or "").strip()
            if cert_id and issuer:
                findings.append(
                    Finding(
                        source_tool=self.name,
                        entity_kind=EntityKind.CERT,
                        entity_value=str(cert_id),
                        attributes={
                            "issuer": issuer,
                            "common_name": (row.get("common_name") or "").strip(),
                            "not_before": row.get("not_before"),
                            "not_after": row.get("not_after"),
                            "domain": target,
                        },
                    )
                )
        return findings

File: lib/tools/crtsh.ts
/**
 * crt.sh — certificate transparency log search.
 *
 * No API key. Returns every certificate ever issued for a domain,
 * which is gold for subdomain enumeration and pivot-by-cert.
 */

import { ToolFunction, type FetchResult, type Finding } from "./base";

export class CrtshTool extends ToolFunction {
  readonly name = "crt_sh";
  readonly accepts = new Set(["domain", "person", "org"]);
  readonly emits = new Set(["cert", "org", "subdomain"]);
  readonly docs_url = "https://crt.sh/";
  readonly description = "Certificate Transparency log search via crt.sh.";
  readonly per_request_timeout_ms = 20_000;

  private static URL = "https://crt.sh/";

  protected async _fetch(query: { target: string }): Promise<FetchResult> {
    const target = (query.target || "").trim().toLowerCase();
    const url = new URL(CrtshTool.URL);
    url.searchParams.set("q", `%.${target}`);
    url.searchParams.set("output", "json");
    const r = await fetch(url.toString(), {
      signal: AbortSignal.timeout(this.per_request_timeout_ms),
    });
    return {
      bytes: Buffer.from(await r.arrayBuffer()),
      contentType: r.headers.get("content-type") || "application/json",
    };
  }

  protected _parse(raw: Buffer, query: { target: string }): Finding[] {
    const target = (query.target || "").trim().toLowerCase();
    let rows: any[] = [];
    try {
      const parsed = JSON.parse(raw.toString("utf-8"));
      if (Array.isArray(parsed)) rows = parsed;
    } catch {
      return [];
    }
    const findings: Finding[] = [];
    const seenSub = new Set<string>();
    for (const row of rows.slice(0, 200)) {
      const name = String(row.name_value || "").trim().toLowerCase();
      if (!name) continue;
      for (const sub of name.split("\n")) {
        const s = sub.trim();
        if (!s.endsWith(target) && s !== target) continue;
        if (seenSub.has(s)) continue;
        seenSub.add(s);
        findings.push({
          source_tool: this.name,
          entity_kind: "subdomain",
          entity_value: s,
          attributes: { parent_domain: target },
        });
      }
      const certId = row.id;
      const issuer = String(row.issuer_name || "").trim();
      if (certId && issuer) {
        findings.push({
          source_tool: this.name,
          entity_kind: "cert",
          entity_value: String(certId),
          attributes: {
            issuer,
            common_name: String(row.common_name || "").trim(),
            not_before: row.not_before,
            not_after: row.not_after,
            domain: target,
          },
        });
      }
    }
    return findings;
  }
}
